File: python/bin/naturalInterpolant.py
##
## This file is part of CNEMLIB.
## 
## Copyright (C) 2003-2011
## Lounes ILLOUL (illoul_lounes@yahoo.fr)
## Philippe LORONG (philippe.lorong@ensam.eu)
## Arts et Metiers ParisTech, Paris, France
## 
## CNEMLIB is free software: you can redistribute it and/or modify
## it under the terms of the GNU Lesser General Public License as published by
## the Free Software Foundation, either version 3 of the License, or
## (at your option) any later version.
##
## CNEMLIB is distributed in the hope that it will be useful,
## but WITHOUT ANY WARRANTY; without even the implied warranty of
## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
## GNU Lesser General Public License for more details.
##
## You should have received a copy of the GNU Lesser General Public License
## along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
##
## Please report bugs to illoul_lounes@yahoo.fr 

import CNEM3D
from numpy import array as nparray
from numpy import zeros
from numpy import roll
from scipy import sparse
import math
import array
import os
import numpy as np

my_dir=os.path.abspath(os.path.dirname(__file__))
#os.putenv('PATH',os.getenv('PATH')+os.pathsep+my_dir)

import sys
sys.path.append('Z:/DATA/1/dev/actu/mesh')

class interpol:

    def __init__(self,XYZ_Noeud,IN_Tri_Ini,XYZ_Point=None,Type_FF=None,Mesher='python'):

        self.point_set=XYZ_Point is not None and Type_FF is not None

        Type_Call_Tet=0
        XYZ_Noeud_c=list(XYZ_Noeud)
        IN_Tri_Ini_c=list(IN_Tri_Ini)

        self.nb_noeud_ini=len(XYZ_Noeud_c)/3

        if Mesher == 'tetgen' :

            if len(IN_Tri_Ini)!=0:
                bbox=file(my_dir+'/bbox_03','rb')
                nb_noeud_tri_bbox=array.array('I')
                nb_noeud_tri_bbox.read(bbox,2)
                noeud_bbox=array.array('f')
                noeud_bbox.read(bbox,nb_noeud_tri_bbox[0]*3)
                in_tri_bbox=array.array('I')
                in_tri_bbox.read(bbox,nb_noeud_tri_bbox[1]*3)
                bbox.close()

                min_max_xyz=[min(XYZ_Noeud_c[0:3*self.nb_noeud_ini:3]),max(XYZ_Noeud_c[0:3*self.nb_noeud_ini:3]),\
                             min(XYZ_Noeud_c[1:3*self.nb_noeud_ini:3]),max(XYZ_Noeud_c[1:3*self.nb_noeud_ini:3]),\
                             min(XYZ_Noeud_c[2:3*self.nb_noeud_ini:3]),max(XYZ_Noeud_c[2:3*self.nb_noeud_ini:3])]
                O=[(min_max_xyz[1]+min_max_xyz[0])/2,\
                   (min_max_xyz[3]+min_max_xyz[2])/2,\
                   (min_max_xyz[5]+min_max_xyz[4])/2]
                R=0.5*math.sqrt(3.)*max([min_max_xyz[1]-min_max_xyz[0],\
                                         min_max_xyz[3]-min_max_xyz[2],\
                                         min_max_xyz[5]-min_max_xyz[4]])
                Coef=1.5
                                                                    
                for i in xrange(len(noeud_bbox)/3) :
                    for j in range(3):
                        XYZ_Noeud_c.append(float(noeud_bbox[3*i+j])*R*Coef+O[j])
                
                for id_node in in_tri_bbox :
                    IN_Tri_Ini_c.append(float(id_node)+self.nb_noeud_ini)

            #IOMaillage.OUT_TRI_V3D([XYZ_Noeud_c,IN_Tri_Ini_c],[[1.,1.,0.,0.5],[0.,1.,0.]],'Tri_V3D_1',1)
            #v3d.show('Tri_V3D_1.tri')

            [IN_New_Old,IN_Old_New,INV_NN,PNV_NN,self.my_id]=\
                CNEM3D.INTERPOL_CNEM3D_ini_tg(XYZ_Noeud_c,IN_Tri_Ini_c,Type_Call_Tet,my_dir+os.path.sep)

            self.INV_NN=list(INV_NN)
            self.PNV_NN=list(PNV_NN)
            
            self.IN_New_Old=nparray(IN_New_Old)-1
            self.IN_Old_New=nparray(IN_Old_New)-1

        else :

            from scipy.spatial import Delaunay
            import Mesh_Util

            DT=Delaunay(nparray(XYZ_Noeud_c).reshape(-1,3))
            #DTri=DT.convex_hull
            #DTet=DT.simplices
            #DNet=DT.neighbors
            
            DTet=DT.simplices.reshape(-1)
            idgTet,volTet=Mesh_Util.Groupement_Tet(DTet,[],XYZ_Noeud_c)
            DNet,DTri,idTetTri=Mesh_Util.LITetV_F_LINTet(DTet.reshape(-1))

            tet_bord=zeros(len(DTet)/4,dtype=np.bool)
            tet_bord[idTetTri]=True
            volTet=np.array(volTet)
            id_tet_flat=np.argwhere(volTet==0.0)

            flat_tet_on_bord=tet_bord[id_tet_flat]

            for i in range(len(volTet)):
                print volTet[i],tet_bord[i]
            
            print id_tet_flat

            print flat_tet_on_bord

            print np.argwhere(flat_tet_on_bord==False)
            
            DNet=roll(nparray(DNet).reshape(-1,4),-1,axis=1).reshape(-1)
            DTri=[float(i+1) for i in DTri]
            DTet=[float(i+1) for i in DTet]
            DNet=[float(i+1) for i in DNet]

            exit(0)

            [self.my_id]=CNEM3D.INTERPOL_CNEM3D_ini_py(XYZ_Noeud_c,DTri,DTet,DNet)
            
            self.INV_NN=[]
            self.PNV_NN=[]

            self.IN_New_Old=nparray(range(self.nb_noeud_ini))
            self.IN_Old_New=nparray(range(self.nb_noeud_ini))            

        if self.point_set : self.set_point(XYZ_Point,Type_FF)

    def set_point(self,XYZ_Point,Type_FF):

        self.point_set=True

        XYZ_Point_c=list(XYZ_Point)
        
        [IN_Point,Nb_Contrib,INV,FF,Grad]=\
         CNEM3D.INTERPOL_CNEM3D_int(self.my_id,XYZ_Point_c,Type_FF)

        INV=list(INV)

        nb_noeud_ini_new=len(self.IN_Old_New)
                
        INV_NN_new=[]
        PNV_NN_new=[]
        new_ind_new_noeud=[-1]*len(self.INV_NN)
        I=0
        for i in range(len(INV)):
            if INV[i]<nb_noeud_ini_new:
                INV[i]=self.IN_Old_New[INV[i]]
            else :
                j=INV[i]-nb_noeud_ini_new
                if new_ind_new_noeud[j]==-1 :
                    I=I+1
                    new_ind_new_noeud[j]=self.nb_noeud_ini+I-1
                    INV_NN_new.extend(self.IN_Old_New[self.INV_NN[3*j:3*(j+1)]])
                    PNV_NN_new.extend(self.PNV_NN[3*j:3*(j+1)])
                INV[i]=new_ind_new_noeud[j]

        INV_NN=nparray(INV_NN_new)
        PNV_NN=nparray(PNV_NN_new)

        if I!=0 :
            indptr=nparray(range(0,INV_NN.shape[0]/3+1),dtype='int')*3        
            self.Mat_INT_NN=sparse.csr_matrix((PNV_NN,INV_NN,indptr),shape=(I,self.nb_noeud_ini),dtype='double')
        else :
            self.Mat_INT_NN=None
    
        IN_Point=nparray(IN_Point)
        inv_IN_Point=zeros(IN_Point.shape,dtype='int')
        inv_IN_Point[IN_Point]=range(0,IN_Point.shape[0])
        
        indptr=zeros(len(Nb_Contrib)+1,dtype='int')
        indptr[1:len(Nb_Contrib)+1]=Nb_Contrib
        for i in xrange(len(Nb_Contrib)):
            indptr[i+1]+=indptr[i]

        Mat_FF=sparse.csr_matrix((FF,INV,indptr),shape=(len(IN_Point),self.nb_noeud_ini+I),dtype='double')
        self.Mat_INT=Mat_FF[inv_IN_Point,:]

        if len(Grad)!=0:
            Mat_GX=sparse.csr_matrix((Grad[::3],INV,indptr),shape=(len(IN_Point),self.nb_noeud_ini+I),dtype='double')
            self.Mat_GX=Mat_GX[inv_IN_Point,:]

            Mat_GY=sparse.csr_matrix((Grad[1::3],INV,indptr),shape=(len(IN_Point),self.nb_noeud_ini+I),dtype='double')
            self.Mat_GY=Mat_GY[inv_IN_Point,:]

            Mat_GZ=sparse.csr_matrix((Grad[2::3],INV,indptr),shape=(len(IN_Point),self.nb_noeud_ini+I),dtype='double')
            self.Mat_GZ=Mat_GZ[inv_IN_Point,:]

        self.In_Out=nparray(Nb_Contrib,dtype='int')
        self.In_Out=self.In_Out[inv_IN_Point]
        self.In_Out=self.In_Out != 0

    def interpolate(self,Var):

        assert self.point_set

        Var_bis=Var
        if self.Mat_INT_NN!=None :
            Var_bis=zeros((Var.shape[0]+self.Mat_INT_NN.shape[0],Var.shape[1]))
            Var_bis[0:Var.shape[0],:]=Var
            Var_bis[Var.shape[0]:Var.shape[0]+self.Mat_INT_NN.shape[0],:]=self.Mat_INT_NN*Var

        return self.Mat_INT*Var_bis

    def mat_interpol_glob(self):
        
        assert self.point_set
        
        if self.Mat_INT_NN!=None :
            Nb_Noeud_Ini=self.Mat_INT_NN.shape[1]
            return self.Mat_INT[:,:Nb_Noeud_Ini]+self.Mat_INT[:,Nb_Noeud_Ini:]*self.Mat_INT_NN
        else :
            return self.Mat_INT

    def __del__(self):

        CNEM3D.INTERPOL_CNEM3D_del(self.my_id)

    @staticmethod
    def in_use():
        return CNEM3D.INTERPOL_CNEM3D_use()
    
def scni(XYZ_Noeud,IN_Tri_Ini,Type_FF,Sup_NN_GS):

    Type_Call_Tet=1
    [IN_New_Old,IN_Old_New,New_Noeud,INV_NN,PNV_NN,\
    Vol_Cel,Nb_Contrib,INV,Grad,IN_Tri,IN_Tet]=\
    CNEM3D.SCNI_CNEM3D(XYZ_Noeud,IN_Tri_Ini,Type_Call_Tet,Type_FF,Sup_NN_GS,my_dir+os.path.sep)

    XYZ_Noeud_New=list()
    for i in IN_Old_New :
        XYZ_Noeud_New.extend(XYZ_Noeud[3*(i-1):3*i])
    XYZ_Noeud_New.extend(New_Noeud)

    IN_Tri_Ini_New=list()
    for i in range(len(IN_Tri_Ini)) :
        IN_Tri_Ini_New.append(IN_New_Old[IN_Tri_Ini[i]]-1)

    XYZ_Noeud_New=nparray(XYZ_Noeud_New)
    XYZ_Noeud_New=XYZ_Noeud_New.reshape((XYZ_Noeud_New.shape[0]/3,3))
    IN_Tri_Ini_New=nparray(IN_Tri_Ini_New)
    IN_Tri_Ini_New=IN_Tri_Ini_New.reshape((IN_Tri_Ini_New.shape[0]/3,3))
    IN_Tri=nparray(IN_Tri)
    IN_Tri=IN_Tri.reshape((IN_Tri.shape[0]/3,3))
    IN_Tet=nparray(IN_Tet)
    IN_Tet=IN_Tet.reshape((IN_Tet.shape[0]/4,4))
    INV_NN=nparray(INV_NN)
    INV_NN=INV_NN.reshape((INV_NN.shape[0]/3,3))
    PNV_NN=nparray(PNV_NN)
    PNV_NN=PNV_NN.reshape((PNV_NN.shape[0]/3,3))
    IN_New_Old=nparray(IN_New_Old)
    IN_Old_New=nparray(IN_Old_New)
    
    return [[Vol_Cel,Nb_Contrib,INV,Grad],\
            XYZ_Noeud_New,IN_Tri_Ini_New,\
            IN_Tri,IN_Tet,\
            INV_NN,PNV_NN,\
            IN_New_Old,IN_Old_New]

def mesh(XYZ_Noeud,IN_Tri_Ini):

    Type_Call_Tet=1
    [IN_New_Old,IN_Old_New,New_Noeud,IN_Tri,IN_Tet]=\
    CNEM3D.MESH_CNEM3D(XYZ_Noeud,IN_Tri_Ini,Type_Call_Tet,my_dir+os.path.sep)

    XYZ_Noeud_New=list()
    for i in IN_Old_New :
        XYZ_Noeud_New.extend(XYZ_Noeud[3*(i-1):3*i])
    XYZ_Noeud_New.extend(New_Noeud)

    IN_Tri_Ini_New=list()
    for i in range(len(IN_Tri_Ini)) :
        IN_Tri_Ini_New.append(IN_New_Old[IN_Tri_Ini[i]]-1)

    XYZ_Noeud_New=nparray(XYZ_Noeud_New)
    XYZ_Noeud_New=XYZ_Noeud_New.reshape((XYZ_Noeud_New.shape[0]/3,3))
    IN_Tri_Ini_New=nparray(IN_Tri_Ini_New)
    IN_Tri_Ini_New=IN_Tri_Ini_New.reshape((IN_Tri_Ini_New.shape[0]/3,3))
    IN_Tri=nparray(IN_Tri)
    IN_Tri=IN_Tri.reshape((IN_Tri.shape[0]/3,3))
    IN_Tet=nparray(IN_Tet)
    IN_Tet=IN_Tet.reshape((IN_Tet.shape[0]/4,4))
    IN_New_Old=nparray(IN_New_Old)
    IN_Old_New=nparray(IN_Old_New)
    
    return [XYZ_Noeud_New,IN_Tri_Ini_New,\
            IN_Tri,IN_Tet,\
            IN_New_Old,IN_Old_New]

File: matlab/bin/naturalInterpolant.m

classdef naturalInterpolant < handle
   properties (Access=public) 
      Points(:,3) double {mustBeReal,mustBeNonNan,mustBeNonsparse}
      Values(:,3) double {mustBeReal,mustBeNonNan,mustBeNonsparse}
      Locations(:,3) double {mustBeReal,mustBeNonNan,mustBeNonsparse}
      Constraints(:,3) double {mustBeReal,mustBeInteger,mustBeNonNan,mustBeNonsparse}
      Method {mustBeMember(Method,{'Sibson','Laplace','SibsonRaw'})}='Sibson'
      Mesher {mustBeMember(Mesher,{'tetgen','matlab'})}='matlab'
   end 
   properties (Access=private)
      Points_changed=false
      Values_changed=false
      Constraints_changed=false
      Locations_changed=false
      Method_changed=false
      Mesher_changed=false
   end
   methods
       function obj = set.Points(obj,Points)
           obj.Points_changed=true;
           obj.Points=Points;
       end
       function obj = set.Values(obj,Values)
           obj.Values_changed=true;
           obj.Values=Values;
       end
       function obj = set.Locations(obj,Locations)
           obj.Locations_changed=true;
           obj.Locations=Locations;
       end
       function obj = set.Method(obj,Method)
           obj.Method_changed=true;
           obj.Method=Method;
       end
       function obj = set.Mesher(obj,Mesher)
           obj.Mesher_changed=true;
           obj.Mesher=Mesher;
       end
       function obj = set.Constraints(obj,Constraints)
           obj.Constraints_changed=true;
           obj.Constraints=Constraints;
       end
       
       function obj=naturalInterpolant(Points,Values,Method,Constraints,Mesher)
           if nargin >= 3
               obj.Points=Points;
               obj.Values=Values;           
               obj.Method=Method;
           end
           
           if nargin >=4 
               obj.Constraints=Constraints;
           end
           
           if nargin >=5
               obj.Mesher=Mesher;
           end
       end
       
       function Evaluation=eval(obj,Locations)
           
           if nargin == 2
               obj.Locations=Locations;
           end
               
           remesh=false;
           if obj.Points_changed
              remesh=true;
              obj.Points_changed=false;
           end
           if obj.Mesher_changed
               remesh=true;
               obj.Mesher_changed=false;
           end
           if obj.Constraints_changed
               remesh=true;
               obj.Constraints_changed=false;
           end
           if remesh
               if size(obj.Points,1) == 0
                   warning('Points not set !')
                   return 
               end
               obj.init();
           end
           
           reinterpol=false;
           if obj.Locations_changed
               reinterpol=true;
               obj.Locations_changed=false;
           end
           if obj.Method_changed
               reinterpol=true;
               obj.Method_changed=false;
           end
           if reinterpol
               if size(obj.Locations,1) == 0
                   warning('Locations not set !')
                   return 
               end
               obj.set_point();
           end
           
           obj.interpolate();
           Evaluation=obj.Var_Int;
       end
       
%        function ret=subsref(obj,Locations)
%            if Locations.type()=='()'
%                obj.Locations=Locations.subs{1}
%                return obj.eval()
%            elseif Locations.type()=='.'
%                 % ???
%            end
%        end
	end
%    methods (Attributes)
%        function obj = methodName(obj,arg2,...)
%           ...
%        end
%     end
%     events (Attributes) 
%        EventName
%     end

    properties(Access=public)
        In_Out
        Nb_V
        Var_Int
        Mat_INT
        Mat_INT_NN
        Mat_GradX
        Mat_GradY
        Mat_GradZ
    end
    properties(Access=private)
        
        nb_noeud_ini
        
        IN_New_Old
        IN_Old_New
        INV_NN
        PNV_NN
        
        my_id
        
        point_set
        
        Type_Call_Tet=0
        
        Type_FF=containers.Map({'Sibson','Laplace','TetLinear','SibsonRaw'},{0,1,2,3})
        
    end
    methods (Access=private)
        function init(obj)
            
            XYZ_Noeud=obj.Points;
            IN_Tri_Ini=obj.Constraints;
            
            obj.nb_noeud_ini=size(XYZ_Noeud,1);
            obj.point_set=false;
            obj.my_id=0;
            
            if obj.Mesher=='tetgen' 
                
                if size(IN_Tri_Ini,1)~=0
                    fid=fopen('bbox_03','r');
                    nb_node_bbox=fread(fid,1,'uint32');
                    nb_tri_bbox=double(fread(fid,1,'uint32'));
                    xyz_node_bbox=fread(fid,nb_node_bbox*3,'float32');
                    in_tri_bbox=fread(fid,nb_tri_bbox*3,'uint32');
                    fclose(fid);
                    xyz_node_bbox=reshape(xyz_node_bbox,3,nb_node_bbox)';
                    in_tri_bbox=reshape(in_tri_bbox,3,nb_tri_bbox)'+1;

                    max_XYZ_Noeud=max(XYZ_Noeud,[],1);
                    min_XYZ_Noeud=min(XYZ_Noeud,[],1);
                    O_dbox=(max_XYZ_Noeud+min_XYZ_Noeud)/2;
                    L_dbox=max(max_XYZ_Noeud-min_XYZ_Noeud);
                    R_dbox=sqrt(3)*L_dbox/2.;
                    coef=1.5;

                    xyz_node_bbox=xyz_node_bbox*(coef*R_dbox)+repmat(O_dbox,size(xyz_node_bbox,1),1);
                    IN_Tri_Ini=[IN_Tri_Ini;in_tri_bbox+size(XYZ_Noeud,1)];
                    XYZ_Noeud=[XYZ_Noeud;xyz_node_bbox];
                end

                %a=sparse(1,1);s=whos('a');
                %if s.bytes==20 %32bit
                %   IN_Tri_Ini=uint32(IN_Tri_Ini);
                %else %64bit
                %   IN_Tri_Ini=uint64(IN_Tri_Ini);
                %end
                
                %S = dbstack('-completenames');
                %my_path=S(1).file;
                [folder, name, ext] = fileparts(which('cnem3d'));
                my_dir=[folder filesep];

                [IN_New_Old,IN_Old_New,INV_NN,PNV_NN,my_id]=...
                cnem3d(2,XYZ_Noeud',IN_Tri_Ini'-1,obj.Type_Call_Tet,my_dir);

                obj.IN_New_Old=double(IN_New_Old)';
                obj.IN_Old_New=double(IN_Old_New)';
                obj.INV_NN=double(INV_NN)'+1;
                obj.PNV_NN=PNV_NN';
                obj.my_id=my_id;
            else
                DT = delaunayTriangulation(XYZ_Noeud(:,1),XYZ_Noeud(:,2),XYZ_Noeud(:,3));
                T=DT.ConnectivityList;
                %T=circshift(T,1,2);
                N=DT.neighbors;
                %N=circshift(N,2,2);
                N(isnan(N)) = 0;
                F=DT.convexHull;
                %DT.Points
                
                [my_id]=...
                cnem3d(3,XYZ_Noeud',F'-1,T'-1,N'-1);
                
                obj.my_id=my_id;
                obj.IN_New_Old=(1:obj.nb_noeud_ini)';
                obj.IN_Old_New=(1:obj.nb_noeud_ini)';
                obj.INV_NN=zeros(0,3);
                obj.PNV_NN=zeros(0,3);
            end
        end
        
        function set_point(obj)
                
            XYZ_Point=obj.Locations;
            Type_FF=obj.Type_FF(obj.Method);
            
            obj.point_set=true;
            
            [Ind_Point,Nb_V,Ind_V,FF,Grad]=...
            cnem3d(4,obj.my_id,XYZ_Point',Type_FF);

            Ind_Point=double(Ind_Point)'+1;
            Nb_V=double(Nb_V)';
            Ind_V=double(Ind_V)'+1;

            nb_noeud_ini_new=size(obj.IN_Old_New,1);

            new_ind_new_noeud=zeros(size(obj.INV_NN,1),1);
            nb_new_noeud=size(obj.INV_NN,1);
            INV_NN_new=zeros(nb_new_noeud,3);
            PNV_NN_new=zeros(nb_new_noeud,3);
            I=0;
            for i=1:size(Ind_V,1)
                if Ind_V(i)<=nb_noeud_ini_new
                    Ind_V(i)=obj.IN_Old_New(Ind_V(i));
                else
                    j=Ind_V(i)-nb_noeud_ini_new;
                    if new_ind_new_noeud(j)==0
                        I=I+1;
                        new_ind_new_noeud(j)=obj.nb_noeud_ini+I;
                        INV_NN_new(I,:)=obj.IN_Old_New(obj.INV_NN(j,:))';
                        PNV_NN_new(I,:)=obj.PNV_NN(j,:);
                    end
                    Ind_V(i)=new_ind_new_noeud(j);
                end
            end
            INV_NN=INV_NN_new(1:I,:);
            PNV_NN=PNV_NN_new(1:I,:);

            INT_NN=[];
            if I~=0
                i_NN=reshape(repmat(1:I,3,1),3*I,1);
                INV_NN=reshape(INV_NN',3*I,1);
                PNV_NN=reshape(PNV_NN',3*I,1);
                INT_NN=sparse(i_NN,INV_NN,PNV_NN,I,obj.nb_noeud_ini);
            end

            Inv_Ind_Point=(1:size(Ind_Point,1));
            Inv_Ind_Point(Ind_Point)=Inv_Ind_Point;

            i_Point=zeros(1,size(Ind_V,1));
            k=1;
            for i=1:size(Nb_V,1)
                for j=1:Nb_V(i)
                    i_Point(k)=i;
                    k=k+1;
                end
            end

            Mat_INT=sparse(i_Point',Ind_V,FF',size(XYZ_Point,1),obj.nb_noeud_ini+I);
            obj.Mat_INT=Mat_INT(Inv_Ind_Point,:);
            
            if size(Grad,2)~=0
                Mat_GradX=sparse(i_Point',Ind_V,Grad(1,:)',size(XYZ_Point,1),obj.nb_noeud_ini+I);
                obj.Mat_GradX=Mat_GradX(Inv_Ind_Point,:);

                Mat_GradY=sparse(i_Point',Ind_V,Grad(2,:)',size(XYZ_Point,1),obj.nb_noeud_ini+I);
                obj.Mat_GradY=Mat_GradY(Inv_Ind_Point,:);

                Mat_GradZ=sparse(i_Point',Ind_V,Grad(3,:)',size(XYZ_Point,1),obj.nb_noeud_ini+I);
                obj.Mat_GradZ=Mat_GradZ(Inv_Ind_Point,:);
            end

            Nb_V(Ind_Point)=Nb_V;
            obj.Nb_V=double(Nb_V);
            obj.In_Out=logical(Nb_V);
            
            obj.Mat_INT_NN=INT_NN;
        end
        
        function delete(obj)
            if obj.my_id ~= 0   
                cnem3d(5,obj.my_id);
            end
        end
        
        function interpolate(obj)
            Var=obj.Values;
            Var_Add_Node=[];
            if size(obj.Mat_INT_NN,1)~=0
                Var_Add_Node=obj.Mat_INT_NN*Var;
            end 
            obj.Var_Int=obj.Mat_INT*[Var;Var_Add_Node];
        end
        
        function Mat=mat_interpol_glob(obj)
            if size(obj.Mat_INT_NN,1)~=0
                n=size(obj.Mat_INT_NN,2);
                m=size(obj.Mat_INT,2);
                Mat=obj.Mat_INT(:,1:n)+obj.Mat_INT(:,n+1:m)*obj.Mat_INT_NN;
            else
                Mat=obj.Mat_INT;
            end
        end
    end
    methods(Static)
        function in_use()
            cnem3d(6);
        end 
    end
end

       

       
       